lib/rspec/core/formatters.rb

```diff
@@ -143,9 +143,11 @@ def add(formatter_to_use, *paths)
 
       if !Loader.formatters[formatter_class].nil?
         formatter = formatter_class.new(*args)
+        return formatter if duplicate_formatter_exists?(formatter)
         @reporter.register_listener formatter, *notifications_for(formatter_class)
       elsif defined?(RSpec::LegacyFormatters)
         formatter = RSpec::LegacyFormatters.load_formatter formatter_class, *args
+        return formatter if duplicate_formatter_exists?(formatter)
```

Comment:
You've got two `return` statements here but only one spec.  Presumably only one of them is tested?  Would it be easy to add a test for the other branch?

Alternately (or maybe in addition...), there's duplication here, and it's definitely duplicated knowledge.  Instead of `return formatter if ...; @reporter.register_listener`, WDYT about extracting a private `register_listener` helper method that would do the `duplicate_formatter_exists?` check and then call `@reporter.register_listener`?  I'd feel better about having a spec for only one of these branches if they were going through a shared helper method.